File: app/src/main/java/dev/bltucker/nanodegreecapstone/storydetail/StoryCommentsAdapter.java
package dev.bltucker.nanodegreecapstone.storydetail;

import android.content.res.Resources;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.Calendar;

import javax.inject.Inject;

import dev.bltucker.nanodegreecapstone.R;
import dev.bltucker.nanodegreecapstone.databinding.StoryCommentLayoutItemBinding;
import dev.bltucker.nanodegreecapstone.injection.GregorianUTC;
import dev.bltucker.nanodegreecapstone.models.Comment;

public class StoryCommentsAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    private static final int COMMENT_ITEM_TYPE = 2;

    private final Resources resources;
    private final Calendar calendar;

    private Comment[] comments;

    @Inject
    public StoryCommentsAdapter(@GregorianUTC Calendar utcCalendar, Resources resources) {
        this.resources = resources;
        calendar = utcCalendar;
        comments = new Comment[0];
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        StoryCommentLayoutItemBinding binding = StoryCommentLayoutItemBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
        binding.getRoot().setTag(R.id.comment_container, binding.commentContainer);
        return new CommentViewHolder(binding, calendar, resources);
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        if (comments.length == 0) {
            return;
        }

        Comment comment = comments[position];

        if (null == comment) {
            return;
        }

        CommentViewHolder commentViewHolder = (CommentViewHolder) holder;
        commentViewHolder.bind(comment);
    }

    @Override
    public int getItemViewType(int position) {
        return COMMENT_ITEM_TYPE;
    }

    @Override
    public int getItemCount() {
        return comments.length;
    }

    void updateComments(Comment[] updatedComments) {
        final DetailStoryDiffCallback diffCallback = new DetailStoryDiffCallback(comments, updatedComments);
        final DiffUtil.DiffResult diffResult = DiffUtil.calculateDiff(diffCallback);
        comments = updatedComments;
        diffResult.dispatchUpdatesTo(StoryCommentsAdapter.this);
    }

}

File: app/src/main/java/dev/bltucker/nanodegreecapstone/storydetail/StoryCommentsAdapter.kt
package dev.bltucker.nanodegreecapstone.storydetail

import android.content.res.Resources
import android.support.v7.util.DiffUtil
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.ViewGroup
import dev.bltucker.nanodegreecapstone.R
import dev.bltucker.nanodegreecapstone.databinding.StoryCommentLayoutItemBinding
import dev.bltucker.nanodegreecapstone.injection.GregorianUTC
import dev.bltucker.nanodegreecapstone.models.Comment
import java.util.*
import javax.inject.Inject

class StoryCommentsAdapter @Inject
constructor(@param:GregorianUTC private val calendar: Calendar, private val resources: Resources) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private var comments: Array<Comment>

    init {
        comments = arrayOf()
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val binding = StoryCommentLayoutItemBinding.inflate(LayoutInflater.from(parent.context), parent, false)
        binding.root.setTag(R.id.comment_container, binding.commentContainer)
        return CommentViewHolder(binding, calendar, resources)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (comments.isEmpty()) {
            return
        }

        val comment = comments[position]

        val commentViewHolder = holder as CommentViewHolder
        commentViewHolder.bind(comment)
    }

    override fun getItemViewType(position: Int): Int {
        return COMMENT_ITEM_TYPE
    }

    override fun getItemCount(): Int {
        return comments.size
    }

    internal fun updateComments(updatedComments: Array<Comment>) {
        val diffCallback = DetailStoryDiffCallback(comments, updatedComments)
        val diffResult = DiffUtil.calculateDiff(diffCallback)
        comments = updatedComments
        diffResult.dispatchUpdatesTo(this@StoryCommentsAdapter)
    }

    companion object {
        private val COMMENT_ITEM_TYPE = R.layout.story_comment_layout_item
    }

}
